Support apply PVC template changes to StatefulSet

d-kuro opened the pull request:
refs: https://github.com/cybozu-go/moco/issues/265

~~Add design documentation for moco-controller to support volume size expansion~~

> **Note**
>
> Rewrote docs to focus on recreating the StatefulSet rather than extending volume

ymmt2005 reviewed (dismissed):
LGTM

masa213f commented on docs/designdoc/support_volume_expansion.md:
````diff
@@ -0,0 +1,89 @@
+      message: "Validation failed: ..."
+```
+
+Set the `VolumeResized` condition's status to `True` if the PVC expansion was successful.
````
That fine with me.
Could you also export this status as a metric?

d-kuro replied:
I see, I came up with two options, which would you prefer?

1. Count metric, `moco_pvc_expansion_failed`
    * Count is incremented if PVC resizing fails
2. Gauge metric, `moco_pvc_expansion_status{states="True/False"}`
    * Aggregate the number of objects from the list of MySQLCluster's that currently exist with that status

masa213f replied:
Could you export this metrics for each MySQLCluster?
And also use the prefix of `moco_cluster_` as with other metrics.
https://github.com/cybozu-go/moco/blob/main/docs/metrics.md#mysql-clusters

How about use the following format?
Count metircs: `moco_cluster_volume_resized_total{status="True/False", name="mycluster", namespace="mynamesapce"}`

ymmt2005 replied:
@masa213f 
As I told you elsewhere, moco-controller should re-create StatefulSet not only when the storage size is changed, but also when metadata (labels / annotations) is changed. Considering that, the metric name containing `volume_resized` is not appropriate.

ymmt2005 replied:
@d-kuro Have you heard about that?

d-kuro replied:
I updated the docs to reflect the comments 🙏 :
https://github.com/cybozu-go/moco/pull/412/commits/e63a1d56f2c81684fc0c7492e43058c1538c4735

d-kuro commented:
@ymmt2005 @masa213f 

I have rewritten the document based on conversations we had at our meetings.
Added metrics to notify errors, as moco-controller may endlessly handle errors that are not automatically recovered in some cases.
Please re-review :pray:
https://github.com/cybozu-go/moco/pull/412/commits/11475fa8d5bdad5689fc82cff3a5e0011a05b3c6

masa213f reviewed (approved):
LGTM. Great job! Thank you.

donch commented:
Hello, 

Do you have any release date which will include this feature please ?

masa213f commented:
@donch 
thank you for contacting. I will release it in the next week.

masa213f commented:
Sorry, preparation of release has been delayed. I will release it next week. :(
